libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/incremental/ClasspathChangesComputer.kt

```diff
@@ -5,184 +5,94 @@
 
 /** Computes [ClasspathChanges] between two [ClasspathSnapshot]s .*/
 object ClasspathChangesComputer {
 
-    fun getChanges(current: ClasspathSnapshot, previous: ClasspathSnapshot): ClasspathChanges {
-        val changesCollector = ChangesCollector()
-        return when (collectClasspathChanges(current, previous, changesCollector)) {
-            Success -> {
-                val (lookupSymbols, fqNames, _) = changesCollector.getDirtyData(emptyList(), NoOpBuildReporter)
-                ClasspathChanges.Available(lookupSymbols.toList(), fqNames.toList())
-            }
-            is Failure -> ClasspathChanges.NotAvailable.UnableToCompute
-        }
-    }
-
-    private fun collectClasspathChanges(
-        current: ClasspathSnapshot,
-        previous: ClasspathSnapshot,
-        changesCollector: ChangesCollector
-    ): ChangesCollectorResult {
-        if (current.classpathEntrySnapshots.size != previous.classpathEntrySnapshots.size) {
-            return Failure.AddedRemovedClasspathEntries
-        }
+    fun compute(currentClasspathSnapshot: ClasspathSnapshot, previousClasspathSnapshot: ClasspathSnapshot): ClasspathChanges {
+        val currentClassSnapshots = currentClasspathSnapshot.getClassSnapshots()
+        val previousClassSnapshots = previousClasspathSnapshot.getClassSnapshots()
 
-        for (index in current.classpathEntrySnapshots.indices) {
-            val result = collectClasspathEntryChanges(
-                current.classpathEntrySnapshots[index],
-                previous.classpathEntrySnapshots[index],
-                changesCollector
-            )
-            if (result is Failure) {
-                return result
-            }
-        }
-        return Success
-    }
-
-    private fun collectClasspathEntryChanges(
-        current: ClasspathEntrySnapshot,
-        previous: ClasspathEntrySnapshot,
-        changesCollector: ChangesCollector
-    ): ChangesCollectorResult {
-        if (current.classSnapshots.size != previous.classSnapshots.size) {
-            return Failure.AddedRemovedClasses
-        }
+        val workingDir =
+            FileUtil.createTempDirectory(this::class.java.simpleName, "_WorkingDir_${UUID.randomUUID()}", /* deleteOnExit */ true)
+        val incrementalJvmCache = IncrementalJvmCache(workingDir, /* targetOutputDir */ null, FileToCanonicalPathConverter)
 
-        for (key in current.classSnapshots.keys) {
-            val currentSnapshot = current.classSnapshots[key]!!
-            val previousSnapshot = previous.classSnapshots[key] ?: return Failure.AddedRemovedClasses
-            val result = collectClassChanges(currentSnapshot, previousSnapshot, changesCollector)
-            if (result !is Success) {
-                return result
+        // Store previous class snapshots in incrementalJvmCache, the returned ChangesCollector result is not used.
+        val unusedChangesCollector = ChangesCollector()
+        for (previousSnapshot in previousClassSnapshots) {
```

Comment:
(next PR) Can you please add more performance-related logging, using `info` level is probably ok? It'd be interesting to know how much time is spent in: computing individual ABI snapshots, loading previous, loading current, diffing the two etc.

Reply:
Yes, once we've guaranteed correctness, I'll start measuring & optimizing.